Add vm.unixTime() cheatcode

Tudmotu opened the pull request:
## Motivation
Fixes #5937.
For long-running scripts it is sometimes useful to measure certain parts and analyze their performance in order to optimize running time. `block.timestamp` does not provide a real-time system timestamp, and using `date` via `ffi` is not granular enough due to Mac OS lack of milliseconds support.

This feature will return an accurate timestamp, in milliseconds since unix epoch, as `uint`.

## Solution
I've added the usual cheatcode components:
- Added function to `HEVM.sol`
- Generated abi bindings
- Added implementation to `foundry-evm` (specifically, `ext.rs`)
- Added Solidity tests in `testdata`

Implementation details:
- Using `SystemTime::now().duration_since(UNIX_EPOCH)`, following the [example from the docs](https://doc.rust-lang.org/std/time/struct.SystemTime.html#impl-SystemTime)
- Converting `u128` to `Bytes` via `rU256` 

## Open questions
- The following test fails both on my branch and on master: `'otterscan::can_call_ots_has_code'`.
Is this ok?
- Using`rU256` to convert the `u128` from `.as_millis()` to bytes — is there a nicer way to achieve this?
- Should this be renamed to something else, such as `.unixTime()`, `.systemTime()` or `.timestamp()`? Personally I would have preferred `vm.now()` but `now` is a reserved word in Solidity

mattsse commented on crates/evm/src/executor/inspector/cheatcodes/ext.rs:
```diff
@@ -511,6 +517,16 @@ fn sleep(milliseconds: &U256) -> Result {
 }
 
+/// Returns the time since unix epoch in milliseconds
+fn time() -> Result {
```
let's call this duration_since_epoch

mattsse commented on crates/evm/src/executor/inspector/cheatcodes/ext.rs:
```diff
@@ -511,6 +517,16 @@ fn sleep(milliseconds: &U256) -> Result {
+    let difference =
+        sys_time.duration_since(UNIX_EPOCH).expect("Failed getting timestamp in time cheatcode");
+    let millis = difference.as_millis();
+    let as_u8_arr: [u8; 32] = rU256::from(millis).to_be_bytes();
+    Ok(Bytes::from(as_u8_arr))
```
we need to encode this, types are currently a bit horrible until we completely transitioned to alloy types we need to use:
`Ok(rU256::from(millis).to_ethers().encode().into())`

mattsse reviewed (changes requested):
cool, I like this,

smol nits

Tudmotu commented:
Thanks @mattsse for the review, will make the changes and update :slightly_smiling_face:

Tudmotu commented:
@mattsse I have addressed your comments and also went ahead and changed the cheatcode to `unixTime()`.
Can still modify it if there's any objection to that name.

Evalir reviewed (approved):
personally don't have typos—as long as this test passes I think this is good! deferring to @mattsse & @mds1 for merge (also needs `cargo +nightly fmt`)

Tudmotu commented:
Apologies, I messed up the format in the last change and forgot to fix it.
Pushed a fix now.

Evalir reviewed (approved):
let's remember to update the book & forge-std once this is merged!

Tudmotu commented:
@Evalir I already got foundry book & forge-std updates prepared, just waiting for this to be merged before I open PRs :slightly_smiling_face:

mattsse reviewed (approved):
lgtm, tysm!

Tudmotu commented:
Thank you all, you're doing amazing work with Foundry. And I must add the PR process was very smooth, well done! :clap:
